# AI 534 Homework 4 {-}

Author: Tanner Wells

## Part 1: {-}

1.1 

In [246]:
#Imports
import pandas as pd
import numpy as np

from gensim.models import KeyedVectors
from sklearn.metrics.pairwise import cosine_similarity
import train
from svector import svector

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

from sklearn.linear_model import Perceptron

from sklearn.preprocessing import OneHotEncoder

In [247]:
wv = KeyedVectors.load('embs_train.kv')


1.2

a. 

In [248]:
num_sim = 10

wonderful = wv.most_similar('wonderful', topn=num_sim)
awful = wv.most_similar('awful', topn=num_sim)

print("Most similar words to `wonderful`:")
for i in range(num_sim):
    print(wonderful[i][0])

print("\nMost similar words to `awful`:")
for i in range(num_sim):
    print(awful[i][0])

Most similar words to `wonderful`:
marvelous
fantastic
great
fabulous
terrific
lovely
amazing
beautiful
magnificent
delightful

Most similar words to `awful`:
horrible
terrible
dreadful
horrid
atrocious
ugly
lousy
unbelievable
appalling
hideous


The code output above shows the top ten most similar words to "wonderful" and "awful". These results make a lot of sense. The words similar to "wonderful" are very positive and the words similar to "awful" are very negative

b.

In [249]:
american = wv.most_similar('american', topn=num_sim)
twelve = wv.most_similar('twelve', topn=num_sim)
key = wv.most_similar('key', topn=num_sim)

print("Most similar words to `american`:")
for i in range(num_sim):
    print(american[i][0])

print("\nMost similar words to `twelve`:")
for i in range(num_sim):
    print(twelve[i][0])

print("\nMost similar words to `key`:")
for i in range(num_sim):
    print(key[i][0])

Most similar words to `american`:
america
americans
british
indian
canadian
chinese
european
mexican
african
australian

Most similar words to `twelve`:
thirteen
fourteen
fifteen
eleven
ten
twenty
thirty
six
seventy
five

Most similar words to `key`:
crucial
vital
critical
important
essential
integral
main
major
paramount
significant


The code output above shows the top ten most similar words to "american", "twelve" and "key". These results make a lot of sense. The similar words to "american" are other nationalities, the similar words to "twelve" are numbers, and similar words to "key" are synonyms.

1.3

In [250]:
brother = wv.most_similar(positive=['sister', 'man'], negative=['woman'], topn=num_sim)
faster = wv.most_similar(positive=['harder', 'fast'], negative=['hard'], topn=num_sim)

print("Most similar words query 1:")
for i in range(num_sim):
    print(brother[i][0])

print("\nMost similar words query 2:")
for i in range(num_sim):
    print(faster[i][0])

Most similar words query 1:
brother
uncle
nephew
son
father
brothers
dad
siblings
daughter
sons

Most similar words query 2:
faster
rapidly
easier
slow
quickly
bigger
cheaper
louder
slowly
smarter


The code output above shows the top ten most similar words for "sister - woman + man" and "harder - hard + fast". These results make a lot of sense. The top output for the first query is "brother", which would be my first guess. The same is true for the second query, whose first output is "faster".

In [251]:
one = wv.most_similar(positive=['key', 'object'], negative=['important'], topn=num_sim)
two = wv.most_similar(positive=['america', 'state'], negative=['country'], topn=num_sim)
three = wv.most_similar(positive=['college', 'group'], negative=['academic'], topn=num_sim)

print("Most similar words query 1:")
for i in range(num_sim):
    print(one[i][0])

print("\nMost similar words query 2:")
for i in range(num_sim):
    print(two[i][0])

print("\nMost similar words query 3:")
for i in range(num_sim):
    print(three[i][0])



Most similar words query 1:
objects
fragment
switchblade
bolt
stick
gravitational
triangle
wrench
shield
beam

Most similar words query 2:
washington
texas
tennessee
penn
alabama
ralph
california
franklin
monroe
arkansas

Most similar words query 3:
groups
who
young
bunch
somethings
younger
trio
association
posse
buddies


The code output above shows the top ten most similar words for "key - important + object", "america - country + state" and "college - academic + group". These results mostly make sense. For the first query, I tried to get objects most similar to a key without words like "paramount", which worked. for the second result, I tired to get a list of American States, which mostly worked except it also seems to have included colleges like "penn" and people like "franklin". For the third query, I tried to get words similar to college without association to academics, which seems to have mostly worked. 

## Part 2: {-}

2.1.1 

In [252]:
#print(wv[0])
vector_list = []
average_list = []
#Use svector to make list of words
for i, (label, words) in enumerate(train.read_from('train.csv'), 1): # label is +1 or -1
    sent = train.make_vector(words)
    vector_list.append(sent)

for idx, list in enumerate(vector_list):
    n = 0
    for word_no,word in enumerate(vector_list[idx]):
        
        try:
            similar = wv[word]
            
            if n == 0:
                similar_list = wv[word]
            else:
                similar_list=wv[word]+similar_list
            n = n+1
        except KeyError:
            pass

    similar_average = similar_list/n
    average_list.append(similar_average)




In [253]:
data = pd.read_csv('train.csv')
    
# Compute cosine similarity between the first embedding and the rest
similarities = cosine_similarity([average_list[0]], average_list[1:])
    
# Find the index of the highest similarity
most_similar_index = np.argmax(similarities)+1
print("First sentence in training set:")
print(data['sentence'][0])
print("\nClosest sentence to first sentence:")
print(data['sentence'][most_similar_index])

First sentence in training set:
it 's a tour de force , written and directed so quietly that it 's implosion rather than explosion you fear

Closest sentence to first sentence:
a semi autobiographical film that 's so sloppily written and cast that you can not believe anyone more central to the creation of bugsy than the caterer had anything to do with it


The code output above shows the sentence most similar to the first sentence in the training set. Honestly, I thought both reviews were negative but apparently the first review is positive. It makes sense that the closest sentence is a negative review, even if the labels do not agree. Both reviews seem to critisize the writing, so in this way the meaning is similar. 

2.1.2

In [254]:
# Compute cosine similarity between the first embedding and the rest
similarities_2 = cosine_similarity([average_list[1]], average_list[:1]+average_list[2:])
    
# Find the index of the highest similarity
most_similar_index_2 = np.argmax(similarities_2)+1
print("Second sentence in training set:")
print(data['sentence'][1])
print("\nClosest sentence to second sentence:")
print(data['sentence'][most_similar_index_2])

Second sentence in training set:
places a slightly believable love triangle in a difficult to swallow setting , and then disappointingly moves the story into the realm of an improbable thriller

Closest sentence to second sentence:
before it takes a sudden turn and devolves into a bizarre sort of romantic comedy , steven shainberg 's adaptation of mary gaitskill 's harrowing short story is a brilliantly played , deeply unsettling experience


The code output above shows the sentence most similar to the second sentence in the training set. Both reviews are clearly negative reviews, so the label makes sense. Both reviews critisize the romantic part of the movie, so it makes sense that the sentence embeddings would be grouped together. 

2.1.3

In [255]:
#Import test data
data_dev = pd.read_csv('dev.csv')
#Copy code from HW2 to encode target
target_dev = []
for sign in data_dev["target"]:
    if sign == "+":
        target_dev.append(1)
    else:
        target_dev.append(0)
y_dev_kNN= np.array(target_dev)





In [256]:
#Set up embeddings on test
vector_list_dev = []
average_list_dev = []
#Use svector to make list of words
for i, (label, words) in enumerate(train.read_from('dev.csv'), 1): # label is +1 or -1
    sent = train.make_vector(words)
    vector_list_dev.append(sent)

for idx, list in enumerate(vector_list_dev):
    n = 0
    for word_no,word in enumerate(vector_list_dev[idx]):
        
        try:
            similar = wv[word]
            
            if n == 0:
                similar_list = wv[word]
            else:
                similar_list=wv[word]+similar_list
            n = n+1
        except KeyError:
            pass
    if n == 0:
        similar_average = similar_list/(word_no+1)
    else:
        similar_average = similar_list/n
    average_list_dev.append(similar_average)

In [257]:
 #Copy code from HW2 to encode target
target_train = []
err_set_1 = []
for sign in data["target"]:
    if sign == "+":
        target_train.append(1)
    else:
        target_train.append(0)
y_train_kNN= np.array(target_train)

#Copy HW 1 code 
best_dev_error_kNN = 1
print("Running k-NN using sentence embedding")
for i in range (50):
    k = 2*i+1
    neigh = KNeighborsClassifier(n_neighbors=k)
    neigh.fit(average_list, y_train_kNN)


    #Predict and accuracy score
    y_pred_kNN = neigh.predict(average_list_dev)
    dev_error_kNN = 1- accuracy_score(y_dev_kNN,y_pred_kNN) 
    err_set_1.append(dev_error_kNN)
    
    print("k= {0:1d}  dev_err {1:.2%}".format(k,dev_error_kNN))
best_dev_error_kNN = min(err_set_1)
best_k_1 = 2*np.argmin(err_set_1)+1
print("\nBest dev error k-NN sentence embedding: {0:.2%} with k={1:1d}".format(best_dev_error_kNN,best_k_1))

Running k-NN using sentence embedding
k= 1  dev_err 37.20%
k= 3  dev_err 35.00%
k= 5  dev_err 34.50%
k= 7  dev_err 32.50%
k= 9  dev_err 32.40%
k= 11  dev_err 32.00%
k= 13  dev_err 31.50%
k= 15  dev_err 31.10%
k= 17  dev_err 30.40%
k= 19  dev_err 30.50%
k= 21  dev_err 29.80%
k= 23  dev_err 30.40%
k= 25  dev_err 30.90%
k= 27  dev_err 30.90%
k= 29  dev_err 30.80%
k= 31  dev_err 31.00%
k= 33  dev_err 30.40%
k= 35  dev_err 30.10%
k= 37  dev_err 31.40%
k= 39  dev_err 30.50%
k= 41  dev_err 29.50%
k= 43  dev_err 29.10%
k= 45  dev_err 28.80%
k= 47  dev_err 29.00%
k= 49  dev_err 28.80%
k= 51  dev_err 29.90%
k= 53  dev_err 29.20%
k= 55  dev_err 29.30%
k= 57  dev_err 29.10%
k= 59  dev_err 28.70%
k= 61  dev_err 28.80%
k= 63  dev_err 28.70%
k= 65  dev_err 28.40%
k= 67  dev_err 29.10%
k= 69  dev_err 29.30%
k= 71  dev_err 29.10%
k= 73  dev_err 28.90%
k= 75  dev_err 28.90%
k= 77  dev_err 29.30%
k= 79  dev_err 28.70%
k= 81  dev_err 28.90%
k= 83  dev_err 29.20%
k= 85  dev_err 28.80%
k= 87  dev_err 27.80%

2.1.4

In [258]:
#Set up onehotencoder
encoder = OneHotEncoder(handle_unknown='ignore')

sentences_train = data['sentence']
sentences_dev = data_dev['sentence']

words_train = [sentence.split() for sentence in sentences_train]
words_dev = [sentence.split() for sentence in sentences_dev]

all_words_train = [word for sentence in words_train for word in sentence]
words_train_reshaped = np.array(all_words_train).reshape(-1, 1)
encoder.fit(words_train_reshaped)

encoded_sentences_train = [encoder.transform(np.array(sentence).reshape(-1, 1)) for sentence in words_train]


In [259]:
encoded_sentences_dev = [encoder.transform(np.array(sentence).reshape(-1, 1)) for sentence in words_dev]
encoded_sentences_dense_train = [sentence.toarray().sum(axis=0) for sentence in encoded_sentences_train]
encoded_sentences_dense_dev = [sentence.toarray().sum(axis=0) for sentence in encoded_sentences_dev]

best_dev_error_kNN_2 = 1
err_set_2 = []
print("Running k-NN using OneHot encoding")
for i in range (50):
    k = 2*i+1
    neigh = KNeighborsClassifier(n_neighbors=k)
    neigh.fit(encoded_sentences_dense_train, y_train_kNN)


    #Predict and accuracy score
    y_pred_kNN = neigh.predict(encoded_sentences_dense_dev)
    dev_error_kNN = 1- accuracy_score(y_dev_kNN,y_pred_kNN) 
    err_set_2.append(dev_error_kNN)
    
    print("k= {0:1d}  dev_err {1:.2%}".format(k,dev_error_kNN))
best_dev_error_kNN_2 = min(err_set_2)
best_k_2 = 2*np.argmin(err_set_2)+1
print("\nBest dev error k-NN sentence embedding: {0:.2%} with k={1:1d}".format(best_dev_error_kNN_2,best_k_2))

Running k-NN using OneHot encoding
k= 1  dev_err 42.80%
k= 3  dev_err 41.40%
k= 5  dev_err 41.80%
k= 7  dev_err 40.60%
k= 9  dev_err 40.30%
k= 11  dev_err 41.80%
k= 13  dev_err 41.70%
k= 15  dev_err 42.90%
k= 17  dev_err 42.90%
k= 19  dev_err 42.60%
k= 21  dev_err 43.00%
k= 23  dev_err 43.10%
k= 25  dev_err 43.50%
k= 27  dev_err 43.30%
k= 29  dev_err 42.20%
k= 31  dev_err 43.00%
k= 33  dev_err 42.90%
k= 35  dev_err 43.50%
k= 37  dev_err 43.40%
k= 39  dev_err 43.60%
k= 41  dev_err 43.90%
k= 43  dev_err 43.80%
k= 45  dev_err 43.60%
k= 47  dev_err 42.20%
k= 49  dev_err 43.80%
k= 51  dev_err 43.10%
k= 53  dev_err 44.20%
k= 55  dev_err 44.20%
k= 57  dev_err 43.90%
k= 59  dev_err 44.80%
k= 61  dev_err 43.20%
k= 63  dev_err 43.50%
k= 65  dev_err 43.60%
k= 67  dev_err 44.70%
k= 69  dev_err 44.50%
k= 71  dev_err 44.00%
k= 73  dev_err 43.60%
k= 75  dev_err 44.30%
k= 77  dev_err 42.70%
k= 79  dev_err 42.70%
k= 81  dev_err 43.80%
k= 83  dev_err 43.90%
k= 85  dev_err 45.10%
k= 87  dev_err 45.00%
k=

2.1.5

In [260]:
#Set up embeddings on test
vector_list_test = []
average_list_test = []
#Use svector to make list of words
for i, (label, words) in enumerate(train.read_from('test.csv'), 1): # label is +1 or -1
    sent = train.make_vector(words)
    vector_list_test.append(sent)

for idx, list in enumerate(vector_list_test):
    n = 0
    for word_no,word in enumerate(vector_list_test[idx]):
        
        try:
            similar = wv[word]
            
            if n == 0:
                similar_list = wv[word]
            else:
                similar_list=wv[word]+similar_list
            n = n+1
        except KeyError:
            pass
    if n == 0:
        similar_average = similar_list/(word_no+1)
    else:
        similar_average = similar_list/n
    average_list_test.append(similar_average)


#Predict Test, use best k from encoder
neigh = KNeighborsClassifier(n_neighbors=best_k_1)
neigh.fit(average_list, y_train_kNN)

#Predict
y_pred_kNN_test = neigh.predict(average_list_test)

list_test = []
for sign in y_pred_kNN_test:
    if sign == 1:
        list_test.append("+")
    else:
        list_test.append("-")


data_test = pd.read_csv('test.csv')
new_csv = data_test
target_list = np.array(list_test)
target_list.reshape(-1,1)
new_csv["target"] = target_list
new_csv.to_csv('test.predicted.csv')




As shown by the screenshot below, my best k-NN classifier lead to a public error rate of .268, which currently ranks 38th on the Kaggle Leaderboard.

![Im1.png](Im1.png)

2.2.1

The code output above shows the basic perception output from HW2. The best dev error rate was 30.1%, which is 2% higher than the best dev error rate for sentence embedding. This shows that sentence embedding can help decrease the erorr rate.

2.2.2

In [261]:
#Imports
from sklearn.linear_model import Perceptron
import time
from svector import svector

In [262]:
class BasicPerceptron:
    def __init__(self, input_dim, learning_rate=0.1):
        self.weights = np.zeros(input_dim)  # Initialize weights to 0
        self.bias = 0  # Initialize bias to 0
        self.learning_rate = learning_rate
    
    def predict(self, x):
        # Simple step function: If the weighted sum is positive, predict 1, else 0
        linear_output = np.dot(x, self.weights) + self.bias
        return np.where(linear_output >= 0, 1, 0)
    
    def train(self, X, y, dev,y_dev,epochs=10):
        best_dev = 1
        for epoch in range(epochs):
            
            for i in range(len(X)):
                # Make a prediction for the current sample
                prediction = self.predict(X[i])
                # Calculate error
                error = y[i] - prediction
                # Update weights and bias if there is an error (Perceptron learning rule)
                self.weights += self.learning_rate * error * X[i]
                self.bias += self.learning_rate * error
            prediction = self.predict(dev)
            test =1- accuracy_score(y_dev,prediction)
            if test<best_dev:
                best_dev = test
                best_epoch = epoch+1
            print("Epoch {0:1d}: Dev Error: {1:.2%}".format(epoch+1,test))
        return best_dev,best_epoch

In [263]:
input_dim = len(average_list[0])  # Dimension of sentence embeddings
perceptron_basic = BasicPerceptron(input_dim)

# Train the perceptron on the sentence embeddings
print("Starting Basic Perceptron")
best_dev_basic,best_epoch_basic = perceptron_basic.train(average_list, y_train_kNN,average_list_dev,y_dev_kNN, epochs=10)

print("\nBest Dev Error: {0:.2%} at epoch {1:1d}".format(best_dev_basic,best_epoch_basic))




Starting Basic Perceptron
Epoch 1: Dev Error: 31.00%
Epoch 2: Dev Error: 40.00%
Epoch 3: Dev Error: 35.00%
Epoch 4: Dev Error: 35.20%
Epoch 5: Dev Error: 40.60%
Epoch 6: Dev Error: 38.90%
Epoch 7: Dev Error: 32.00%
Epoch 8: Dev Error: 29.90%
Epoch 9: Dev Error: 31.30%
Epoch 10: Dev Error: 31.20%

Best Dev Error: 29.90% at epoch 8


The code output above shows the best dev error rate was 29.9%. This is slightly higher than the best dev error rate of 28.2% in HW2. 

2.2.2

In [264]:
class AveragedPerception:
    def __init__(self, input_dim, learning_rate=.1, averaging_factor=0.045):
        self.weights = np.zeros(input_dim)  # Initialize weights to 0
        self.bias = 0  # Initialize bias to 0
        self.avg_weights = np.zeros(input_dim)  # Initialize averaged weights to 0
        self.avg_bias = 0  # Initialize averaged bias to 0
        self.learning_rate = learning_rate
        self.averaging_factor = averaging_factor  # How much to weigh the current weights in the average

    def predict(self, X):
        # Compute the linear combination for the entire batch
        linear_output = np.dot(X, self.avg_weights) + self.avg_bias  # Use averaged weights for prediction
        # Apply step function to each example in the batch
        return np.where(linear_output >= 0, 1, 0)
    
    def train(self, X, y, dev,y_dev,epochs=10):
        best_dev = 1
        for epoch in range(epochs):
            for i in range(len(X)):
                # Make a prediction for the current sample
                prediction = self.predict(X[i].reshape(1, -1))  # Reshape for single sample
                prediction = prediction[0]  # Extract prediction value
                # Calculate error
                error = y[i] - prediction
                # Update weights and bias if there is an error (Perceptron learning rule)
                self.weights += self.learning_rate * error * X[i]
                self.bias += self.learning_rate * error
                
                # Update the averaged weights and bias
                self.avg_weights = (1 - self.averaging_factor) * self.avg_weights + self.averaging_factor * self.weights
                self.avg_bias = (1 - self.averaging_factor) * self.avg_bias + self.averaging_factor * self.bias
            prediction = self.predict(dev)
            test =1- accuracy_score(y_dev,prediction)
            if test<best_dev:
                best_dev = test
                best_epoch = epoch+1
            print("Epoch {0:1d}: Dev Error: {1:.2%}".format(epoch+1,test))

        return best_dev,best_epoch,prediction

In [265]:
input_dim = len(average_list[0])  # Dimension of sentence embeddings
perceptron_av = AveragedPerception(input_dim,averaging_factor=.00000000001)

# Train the perceptron on the sentence embeddings
print("Starting Averaged Perceptron")
best_dev_av,best_epoch_av,prediction_1 = perceptron_av.train(average_list, y_train_kNN,average_list_dev,y_dev_kNN, epochs=10)

print("\nBest Dev Error: {0:.2%} at epoch {1:1d}".format(best_dev_av,best_epoch_av))

Starting Averaged Perceptron
Epoch 1: Dev Error: 30.50%
Epoch 2: Dev Error: 37.90%
Epoch 3: Dev Error: 26.50%
Epoch 4: Dev Error: 31.60%
Epoch 5: Dev Error: 28.10%
Epoch 6: Dev Error: 29.50%
Epoch 7: Dev Error: 26.10%
Epoch 8: Dev Error: 26.50%
Epoch 9: Dev Error: 25.80%
Epoch 10: Dev Error: 29.30%

Best Dev Error: 25.80% at epoch 9


The code output above shows the best dev error rate was 25.8%. This is almost identical to the best dev error rate of 25.9% in HW2. 

2.2.3

Smart averaging is mainly used to speed up the code. In this case, the dimensionality of the weight vector is quite small (300). Implementing smart averaging therefore has diminishing returns, and in my opinion is not completely necessary. However, it still makes the code marginally faster and is definitely not a bad thing to have. 

2.2.4


In [266]:
#Copy code from HW2 to prune 1 count words
#Once again, I am making a new training file to edit. 
from collections import Counter
import warnings
import train4
warnings.filterwarnings('ignore')
#Read the training file and make a list of all words
all_words = []
data_file = pd.read_csv('train.csv')
for i, (label, words) in enumerate(train4.read_from("train.csv"), 1):
    all_words.append(words)
#model_3 = train4.train4("train.csv","dev.csv",2,Bias = 1)
x = all_words
length = max(map(len, x))
array=np.array([xi+[np.nan]*(length-len(xi)) for xi in x])
#print(array)
flat_array = [item for sublist in array for item in sublist]
counts = {}

for item in flat_array:
    counts[item] = counts.get(item, 0) + 1

result = [[item for item in sublist if counts[item] > 1] for sublist in array]
new_csv = data_file
length = max(map(len, result))
sentence_list=np.array([xi+[np.nan]*(length-len(xi)) for xi in result])
n=0
for i in sentence_list:
    edited_sentence=""
    for w in sentence_list[n]:
        if not w=="nan":
            edited_sentence = edited_sentence+" "+w
    new_csv["sentence"][n] = edited_sentence
    n=n+1
#print(new_csv["sentence"][0])
new_csv.to_csv('train_removed.csv',index=False)

In [267]:
#Now re-set up embeddings with one-count words pruned for train
vector_list_prune = []
average_list_prune = []
#Use svector to make list of words
#for i, (label, words) in enumerate(train.read_from('train_removed.csv'), 1):
for i, (words) in enumerate(new_csv["sentence"]): # label is +1 or -1
    sentence = words.split()
    n = 0
    for word_no,word in enumerate(sentence):
        #print(word)
        try:
            
            similar = wv[word]
            
            if n == 0:
                similar_list_p = wv[word]
            else:
                similar_list_p=wv[word]+similar_list_p
            n = n+1
        except KeyError:
            pass
    if n== 0:
        similar_average_p = similar_list_p/word_no
    else:         
        similar_average_p = similar_list_p/n
    average_list_prune.append(similar_average_p)

In [268]:
#Import test data
data_dev_p = pd.read_csv('dev.csv')
#Copy code from HW2 to encode target
target_dev = []
for sign in data_dev["target"]:
    if sign == "+":
        target_dev.append(1)
    else:
        target_dev.append(0)
y_dev_kNN_p= np.array(target_dev)

In [269]:
#Re-train with one-count words pruned
input_dim = len(average_list_prune[0])  # Dimension of sentence embeddings
perceptron_av_prune = AveragedPerception(input_dim,learning_rate=1,averaging_factor=.0000000001)

# Train the perceptron on the sentence embeddings
print("Starting Averaged Percepron")
best_dev_prune,best_epoch_prune,prediction_best = perceptron_av_prune.train(average_list_prune, y_train_kNN,average_list_dev,y_dev_kNN_p, epochs=10)

print("\nBest Dev Error: {0:.2%} at epoch {1:1d}".format(best_dev_prune,best_epoch_prune))


Starting Averaged Percepron
Epoch 1: Dev Error: 32.60%
Epoch 2: Dev Error: 31.60%
Epoch 3: Dev Error: 34.70%
Epoch 4: Dev Error: 29.80%
Epoch 5: Dev Error: 29.30%
Epoch 6: Dev Error: 33.50%
Epoch 7: Dev Error: 37.00%
Epoch 8: Dev Error: 24.60%
Epoch 9: Dev Error: 35.20%
Epoch 10: Dev Error: 29.10%

Best Dev Error: 24.60% at epoch 8


The code output above shows the best dev error rate was 24.6%. This is about one percent lower than the best dev error rate of 25.7% in HW2. 

2.2.5




In [270]:
#Set up best OneHotencoder

#Predict and accuracy score
input_dim = 15805
perceptron_av_OH = AveragedPerception(input_dim,learning_rate=1,averaging_factor=.0000000001)

best_dev_OH,best_epoch_OH,y_pred_kNN_OH = perceptron_av_OH.train(encoded_sentences_dense_train, y_train_kNN,encoded_sentences_dense_dev,y_dev_kNN_p, epochs=10)
dev_error_kNN_OH = 1- accuracy_score(y_dev_kNN,y_pred_kNN) 

print("\nBest Dev Error: {0:.2%} at epoch {1:1d}".format(best_dev_OH,best_epoch_OH))
#Actual: y_dev_kNN_p
#OneHot: y_pred_kNN_OH
#Perceptron: prediction_best


Epoch 1: Dev Error: 33.00%
Epoch 2: Dev Error: 32.90%
Epoch 3: Dev Error: 31.40%
Epoch 4: Dev Error: 34.40%
Epoch 5: Dev Error: 34.80%
Epoch 6: Dev Error: 34.40%
Epoch 7: Dev Error: 31.20%
Epoch 8: Dev Error: 30.10%
Epoch 9: Dev Error: 32.10%
Epoch 10: Dev Error: 28.40%

Best Dev Error: 28.40% at epoch 10


In [271]:
#Find all instances of better guess
number = 0
test = 3
for idx, pred in enumerate(y_dev_kNN):
    if y_pred_kNN_OH[idx] != pred and prediction_best[idx] == pred:
        
        if number <test+1 and number>=test-1:
            print(pred)
            print("\nSentence where W2V is correct but not OneHotEncoder:")
            print(data_dev_p['sentence'][idx])
            
        else:
            pass
        number+=1

1

Sentence where W2V is correct but not OneHotEncoder:
writer director david jacobson and his star , jeremy renner , have made a remarkable film that explores the monster 's psychology not in order to excuse him but rather to demonstrate that his pathology evolved from human impulses that grew hideously twisted
1

Sentence where W2V is correct but not OneHotEncoder:
you 've already seen city by the sea under a variety of titles , but it 's worth yet another visit


The code output above shows two example sentences where Word2Vec is correct but not the OneHot encoding. This is mainly because word2Vec can capture the context of words in a sentence, while OneHot Encoding only assigns one inflexible weight to a word and adds up each weight. In the first sentence, words like "intelligent", "stud", and "conviction" have a positive connotation, so the OneHot Encoder sees the sentence as psotive. However, the Word2Vec method correctly identifies that the words around these positive words put them in a negative context, and correctly identifies the sentence as negative. The second sentence does not have many overtyl positive or negative words, so the OneHotEncoder is more likely to guess wrong. The Word2Vec correctly picks up on the positive nature of this review despite not having many overtly positive words.   

2.2.6


In [273]:


#Do pruned trainign on test
input_dim = len(average_list_prune[0])  # Dimension of sentence embeddings
perceptron_av_prune = AveragedPerception(input_dim,learning_rate=1,averaging_factor=.0000000001)

best_dev_prune,best_epoch_prune,prediction_best = perceptron_av_prune.train(average_list_prune, y_train_kNN,average_list_dev,y_dev_kNN_p, epochs=8)

pruned_test = perceptron_av_prune.predict(average_list_test)

list_test = []
for sign in pruned_test:
    if sign == 1:
        list_test.append("+")
    else:
        list_test.append("-")


data_test = pd.read_csv('test.csv')
new_csv = data_test
target_list = np.array(list_test)
target_list.reshape(-1,1)
new_csv["target"] = target_list
new_csv.to_csv('test.predicted.csv')

Epoch 1: Dev Error: 32.60%
Epoch 2: Dev Error: 31.60%
Epoch 3: Dev Error: 34.70%
Epoch 4: Dev Error: 29.80%
Epoch 5: Dev Error: 29.30%
Epoch 6: Dev Error: 33.50%
Epoch 7: Dev Error: 37.00%
Epoch 8: Dev Error: 24.60%


In [274]:
#Do unpruned training on test
input_dim = len(average_list_prune[0])  # Dimension of sentence embeddings
perceptron_av = AveragedPerception(input_dim,learning_rate=1,averaging_factor=.0000000001)

best_dev_unprune,best_epoch_unprune,prediction_best = perceptron_av.train(average_list, y_train_kNN,average_list_dev,y_dev_kNN_p, epochs=9)

unpruned_test = perceptron_av.predict(average_list_test)

list_test = []
for sign in unpruned_test:
    if sign == 1:
        list_test.append("+")
    else:
        list_test.append("-")


data_test = pd.read_csv('test.csv')
new_csv = data_test
target_list = np.array(list_test)
target_list.reshape(-1,1)
new_csv["target"] = target_list
new_csv.to_csv('test.predicted.csv')

Epoch 1: Dev Error: 30.50%
Epoch 2: Dev Error: 37.90%
Epoch 3: Dev Error: 26.50%
Epoch 4: Dev Error: 31.60%
Epoch 5: Dev Error: 28.10%
Epoch 6: Dev Error: 29.50%
Epoch 7: Dev Error: 26.10%
Epoch 8: Dev Error: 26.50%
Epoch 9: Dev Error: 25.80%


The screenshot below shows my best public Kaggle score using averaged perceptron. My unpruned results were slightly better than my pruned results. My best error rate was .238, which currently ranks 31st.

![Im2.png](Im2.png)

2.3

Table 1: Error rate summary

|  | one-hot, best dev | embedding, best dev | best kaggle public |
| --- | --- | --- | --- |
| k-NN | .403 | .278 | .268 |
| perceptron | .284 | .246 | .238 |

## Part 3: {-}



In [275]:
#Imports
import xgboost as xgb

In [276]:
#Lets try XGBoost
dtrain = xgb.DMatrix(average_list, label=y_train_kNN)
ddev = xgb.DMatrix(average_list_dev, label=y_dev_kNN)
dtest = xgb.DMatrix(average_list_test)

params = {
    'objective': 'binary:logistic',  # For binary classification
    'max_depth': 6, #6
    'eta': 0.1,  # Learning rate .1   .243
    'eval_metric': 'logloss' #logloss
}

evals = [(dtrain, 'train'), (ddev, 'eval')]
num_round = 1000
bst = xgb.train(params, dtrain, num_round,evals=evals,early_stopping_rounds=10,verbose_eval=False)

y_pred = bst.predict(ddev)
y_pred_binary = (y_pred > 0.5).astype(int)
accuracy = 1- accuracy_score(y_dev_kNN, y_pred_binary)
print(f"XGB dev set accuracy: {accuracy * 100:.2f}%")

y_pred_test = bst.predict(dtest)
y_pred_test_binary = (y_pred_test > 0.5).astype(int)


XGB dev set accuracy: 24.20%


In [277]:
list_test = []
for sign in y_pred_test_binary:
    if sign == 1:
        list_test.append("+")
    else:
        list_test.append("-")


data_test = pd.read_csv('test.csv')
new_csv = data_test
target_list = np.array(list_test)
target_list.reshape(-1,1)
new_csv["target"] = target_list
new_csv.to_csv('test.predicted.csv', index=False)

2.1 

I chose to try XGBoost. I did not have to make many changes to my code at all, I just implemented the XGBoost method based on the online documentation.  

2.2

My error rate was 24.2%, as shown above. This was achieved using XGBoost with a max depth of 6 and a learning rate of 0.1. My running time, according to Jupyter Notebook, was 4.7 seconds. This was without pruning any vocabulary.

2.3

I learned that averaged perception honestly was very close in performance to this other well known algorithm. My best dev rate was almost identical using XGBoost compared to averaged perception. I imagine that XGBoost does better with extremely large data sets, which are more often used in reality. 

## Part 4: {-}

4.1

My error rate was 24.2%, as shown above. This was achieved using XGBoost with a max depth of 6 and a learning rate of 0.1.

4.2 

My best public error rate was .236, as shown in the screenshot below. This currently ranks 30th on the public leaderboard. This was achieved using XGBoost with a max depth of 6 and a learning rate of 0.1.

![Im3.png](Im3.png)

## Part 5: {-}

5.1

I spent approximately 10 hours on this assignment.

5.2

I would rate this assignment as moderately difficult. 

5.3 

I worked mostly alone.

5.4

I feel that I really only understand about 10% of the Deep Learning material. I am not sure how this homework relates to deep learning/neural networks. 

5.5

The last couple assignments I have been about 1-2% behind the "correct" error rates. It is very difficult to see what I am doing wrong. I wish there were more resources available to help check my method so I could see why my code is getting suboptimal error rates.
